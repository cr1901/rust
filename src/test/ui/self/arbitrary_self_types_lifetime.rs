#![feature(arbitrary_self_types)]

use std::pin::Pin;

#[derive(Debug)]
struct Foo;
#[derive(Debug)]
struct Bar<'a>(&'a Foo);

impl std::ops::Deref for Bar<'_> {
    type Target = Foo;
    fn deref(&self) -> &Self::Target {
        &self.0
    }
}

impl Foo {
    fn a(&self) -> Bar<'_> {
        Bar(self)
    }

    fn b(c: &Self) -> Bar<'_> {
        Bar(c)
    }

    fn c(self: Bar<'_>) -> Bar<'_> {
        self
    }

    fn d(e: Bar<'_>) -> Bar<'_> {
        e
    }

    fn e(self: &Self) -> Bar<'_> {
        Bar(self)
    }

    fn f(self: Bar<'_>) -> impl std::fmt::Debug {
        self
        //~^ ERROR cannot infer an appropriate lifetime
    }

    fn g(self: Bar<'_>) -> impl std::fmt::Debug + '_ {
        self
    }
}

impl<'a> Bar<'a> {
    fn a(self: Bar<'a>, f: &Foo) -> (Bar<'a>, &Foo) { (self, f) }
    fn b(self: Self, f: &Foo) -> (Bar<'a>, &Foo) { (self, f) }
    fn c(self: Bar<'a>, f: &Foo) -> (Self, &Foo) { (self, f) }
}

impl Bar<'_> {
    fn e(self: Self, f: &Foo) -> (Self, &Foo) { (self, f) }
}

struct Baz<T: Unpin> {
    field: T,
}

impl<T: Unpin> Baz<T> {
    fn field(self: Pin<&mut Self>) -> Pin<&mut T> {
        let this = Pin::get_mut(self);
        Pin::new(&mut this.field)
    }
}

fn main() {
    let foo = Foo;
    { foo.a() };
    { Foo::b(&foo) };
    { Bar(&foo).c() };
    { Foo::d(Bar(&foo)) };
    { foo.e() };
    { Bar(&foo).f() };
    { Bar(&foo).g() };
    let mut baz = Baz { field: 0u8 };
    { Pin::new(&mut baz).field() };
}
