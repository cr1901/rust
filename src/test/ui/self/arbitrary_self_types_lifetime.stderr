error: cannot infer an appropriate lifetime
  --> $DIR/arbitrary_self_types_lifetime.rs:39:9
   |
LL |     fn f(self: Bar<'_>) -> impl std::fmt::Debug {
   |                            -------------------- this return type evaluates to the `'static` lifetime...
LL |         self
   |         ^^^^ ...but this borrow...
   |
note: ...can't outlive the anonymous lifetime #1 defined on the method body at 38:5
  --> $DIR/arbitrary_self_types_lifetime.rs:38:5
   |
LL | /     fn f(self: Bar<'_>) -> impl std::fmt::Debug {
LL | |         self
LL | |
LL | |     }
   | |_____^
help: you can add a constraint to the return type to make it last less than `'static` and match the anonymous lifetime #1 defined on the method body at 38:5
   |
LL |     fn f(self: Bar<'_>) -> impl std::fmt::Debug + '_ {
   |                            ^^^^^^^^^^^^^^^^^^^^^^^^^

error: aborting due to previous error

